app/controllers/admin/admin_users_controller.rb

```diff
@@ -69,27 +69,33 @@ def update
   def update_next_of_kin
     @user = authorize User.find_by!(login: params[:user_login])
     fnok = @user.fannish_next_of_kin
+    previous_fnok_user_id = fnok&.kin&.id
     kin = User.find_by(login: params[:next_of_kin_name])
     kin_email = params[:next_of_kin_email]
 
     if kin.blank? && kin_email.blank?
-      if fnok.present?
+      if fnok.blank?
+        flash[:notice] = ts("No change to Fannish next of kin.")
+      else
         fnok.destroy
-        @user.create_log_item({
-                                action: ArchiveConfig.ACTION_REMOVE_FNOK,
-                                fnok_user_id: fnok.kin.id,
-                                admin_id: current_admin.id,
-                                note: "Change made by #{current_admin.login}"
-                              })
+        log_next_of_kin_removed(previous_fnok_user_id)
         flash[:notice] = ts("Fannish next of kin was removed.")
       end
+
+      redirect_to admin_user_path(@user)
+      return
+    end
+
+    if fnok&.kin == kin && fnok&.kin_email == kin_email
+      flash[:notice] = ts("No change to Fannish next of kin.")
       redirect_to admin_user_path(@user)
       return
     end
 
     fnok = @user.build_fannish_next_of_kin if fnok.blank?
     fnok.assign_attributes(kin: kin, kin_email: kin_email)
```

Comment:
Second shot at improving the update logic.

```suggestion
  def update_next_of_kin
    @user = authorize User.find_by!(login: params[:user_login])
    kin = User.find_by(login: params[:next_of_kin_name])
    kin_email = params[:next_of_kin_email]

    fnok = @user.fannish_next_of_kin
    previous_fnok_user_id = fnok&.kin&.id
    fnok ||= @user.build_fannish_next_of_kin
    fnok.assign_attributes(kin: kin, kin_email: kin_email)

    unless fnok.changed?
      flash[:notice] = ts("No change to fannish next of kin.")
      redirect_to admin_user_path(@user)
      return
    end

    # Remove FNOK that already exists.
    if fnok.persisted? && kin.blank? && kin_email.blank?
      fnok.destroy
      log_next_of_kin_removed(previous_fnok_user_id)
      flash[:notice] = ts("Fannish next of kin was removed.")
      redirect_to admin_user_path(@user)
      return
    end
```

This code would require a test for invalid updates to be changed. As the notice "No change to fannish next of kin." will not be shown, but rather the error message from the validation, "Kin email can't be blank"

```
      it "does nothing if fnok is kept undefined" do
        kin = build(:fannish_next_of_kin, user: user, kin_email: "")

        expect(kin).not_to be_valid
        expect(kin.errors[:kin_email]).to include("can't be blank")
        expect(user.reload.log_items).to be_empty
      end
```